### ISSUE TITLE
bug: CSS 'outline' is missing when WebChat buttons, text fields and links receive focus [accessibility]

### ISSUE THREAD

**Issue by nfreear** [microsoft/BotFramework-WebChat#3657]
Hi,

 * Relates to :~ #3655;

## Version

 * WebChat.js version :~ 4.10.1;
     - JS :~ https://cdn.botframework.com/botframework-webchat/4.10.1/webchat.js;
 * All browsers, all browser-versions, all platforms;

## Describe the bug

When I as a visual user navigate a _WebChat_-based Chat-bot using the keyboard, I can't tell which button or text-field has received focus, as the CSS `outline` has been set to `0` or `none`.

 * [Additional explanation on MozDev](https://developer.mozilla.org/en-US/docs/Web/CSS/outline#accessibility_concerns);

## Steps to reproduce

1. Click on '....'
1. Scroll down to '....'
1. See error

## Expected behavior

I expect to see an outline (defaults to blue in some browsers) around the focussed element -- button, text field, link, so I know where I am.

## Potential fix

I propose a new property in [`defaultStyleOptions.js`](https://github.com/microsoft/BotFramework-WebChat/blob/v4.11.0/packages/component/src/Styles/defaultStyleOptions.js) named for example, `outlineOn`. It would be used in the various components, for example:

 * https://github.com/microsoft/BotFramework-WebChat/blob/v4.11.0/packages/component/src/Styles/StyleSet/SendBoxButton.js#L14
 * ...?

For example:

```js
return {
    '&.webchat__icon-button': {
      backgroundColor: 'Transparent',
      border: 0,
      height: '100%',
      outline: outlineOn ? 'initial' : 'none', // Proposed fix!
      padding: 0,
```

[Bug]

**Comment by gabog** [comment]
Hi @compulim, can you review this issue?

**Comment by compulim** [comment]
You could deploy idiosyncratic styling for adding `outline` CSS property. I believe there will be multiple components need to have `outline` added.

You can look at the samples here, https://github.com/microsoft/BotFramework-WebChat/tree/master/samples/02.branding-styling-and-customization/b.idiosyncratic-manual-styles.

Currently, our design team prefer to use a darker color for focus indicator, instead of border-like outlines.

**Comment by corinagum** [comment]
Changing this to feature request - this is by design, and has a workaround as linked above.

This is what Web Chat looks like right now for a focused button:
![image](https://user-images.githubusercontent.com/14900841/104252677-68953a00-5427-11eb-8eed-c66ddfe951f7.png)


Our design does not use outline styling.

**Comment by nfreear** [comment]
Hi @compulim and @corinagum,

Thanks for your responses ... I've gone back to double-check, with this minimal example — no outline, border, `box-shadow` or similar styling on focus (currently on v1.11.0 of `WebChat`):

 * https://microsoft.github.io/BotFramework-WebChat/01.getting-started/b.minimal-bundle/

> You could deploy idiosyncratic styling for adding outline CSS property...
>
In our implementation I've used CSS `box-shadow` — however, I know the importance of a focus indicator to keyboard accessibility, and the average implementor may not ...

> Currently, our design team prefer to use a darker color for focus indicator, instead of border-like outlines.
>
However, that styling isn't on by default. I'm concerned with sensible defaults so that when an average implementor plonks a `WebChat`-based Chat-bot on their site, they don't fall into an accessibility elephant hole.

If I were to do a pull request for the proposed fix above, are you likely to accept it please? That is where I'd like to gauge your interest/ acceptance.

I hope this helps,


Nick

**Comment by corinagum** [comment]
**Please provide repro steps for this issue, which the original issue template asks for. If you can't provide repro steps, then please attach screenshots.**

I am unable to determine what you mean by 'no outline, border, box-shadow or similar styling on focus (currently on v1.11.0 of WebChat):' My screencap above shows you an instance of a button being highlighted on focus. This is default behavior. We still do not have an instance provided by you of no focus shown whatsoever.

- What do you mean by v1.11.0 of Web Chat? Our latest released version is 4.11, is that what you mean?
- Do you mean the sendbox doesn't show focus? Again, this is customizable via styling.
- Do you mean selected activities are not showing focus? Use this sample: https://github.com/microsoft/BotFramework-WebChat/tree/master/samples/04.api/f.selectable-activity

We're always happy to review pull requests created by the community. Please follow the directions in our contributing guidelines and be sure to fill out all requested PR information when filing. Accessibility is extremely high priority for our team and we're committed to making the out-of-the-box experience as accessible as possible.

**Comment by nfreear** [comment]
Hi @corinagum,

Sorry, I meant version '4.11.0', and yes I appreciate a screenshot would help!

See the attached screenshot -- focus is on the "Upload" paperclip button, but as the `outline` has been removed, it is almost impossible to tell (the paperclip becomes slightly darker on focus -- easily missed if you don't concentrate hard, or have lower vision).

 * https://microsoft.github.io/BotFramework-WebChat/02.branding-styling-and-customization/b.idiosyncratic-manual-styles/

(In this example, `outline` is not removed on the adaptive-card buttons that appear if you type "_Help_", so that appears not to be an issue.)

Switching off the `outline` as WebChat does on the key Upload and Send buttons and the Chat input box, without providing a default replacement is bad practice.

> Again, this is customizable via styling.
>
Which misses my point -- the developer needs to know that there is no focus indicator, and why this is important from an accessibility point of view. It makes it too easy to leave the accessibility problem.

I'm proposing that `WebChat` provides the `styleOptions` configuration so that the developer can consciously switch off the `outline`, and implement their own styling should they wish. But that something reasonable and accessible is used by default.

![webchat-idiosyncratic--no-outline-bug_2](https://user-images.githubusercontent.com/131235/104445595-2daf1580-5591-11eb-9b92-c00d7dcf78d7.png)

I hope this helps you understand the issue that I'm raising.

Yours,

Nick

**Comment by nfreear** [comment]
Hi @corinagum,

I acknowledge your frustration that I was slow to provide a screenshot and to clarify what I meant. In future, I'll add the "How to reproduce" steps and/or a screenshot at the outset!

> Currently, our design team prefer to use a darker color for focus indicator, instead of border-like outlines.
>
Ah, I didn't fully understand what you meant until just now ... that is the "_slightly darker on focus_" that I refer to above, and which I argue is too subtle and an insufficient replacement for some sort of border-like outline.


Nick

**Comment by corinagum** [comment]
Thank you for providing the relevant information. I will bring this to our team's attention. However, we have yearly accessibility audits, and this has not been brought up as a deficiency. If we are violating MAS rules, then yes this would become a work item.

In the meantime, manual styling is still an option to our customers to add outline, highlight, etc, if an immediate solution is needed.

@Amit8527510735, could you take a look and let us know if this is an a11y violation?

To test:
1. https://microsoft.github.io/BotFramework-WebChat/01.getting-started/a.full-bundle/
2. Use keyboard to focus on the sendbox buttons (paperclip and/or paper airplane)
3. Note that focus indicator is a darker gray than non-focus. There is no highlight or border indication for focus.

Thank you!

**Comment by nfreear** [comment]
Hi @corinagum and @Amit8527510735,

@corinagum, thanks for your response.

I have not been able to find a publicly available copy of the _Microsoft Accessibility Standard_ ("_MAS_") online to review.

I have gone back and reviewed the following related to WCAG 2.1:

 * [WCAG 2.1 - Success Criterion 2.4.7 Focus Visible](https://w3.org/TR/WCAG21/#focus-visible);
 * [Quick reference for SC 2.4.7][qref-2.4.7];
 * [F78 : Failure of Success Criterion 2.4.7 due to styling element outlines and borders in a way that removes or renders non-visible the visual focus indicator][F78];
 * [G165: Using the default focus indicator for the platform so that high visibility default focus indicators will carry over][G165];
 * [G195: Using an author-supplied, highly visible focus indicator][G195];

My conclusion is that the default behavior is a fairly cut and dried case of failure _F78_.

Going back to my original report, note that I am only proposing a new option `outlineOn` or `focusOutline` or similar, to make it easier for implementors to maintain the default focus indicator, or to do there own thing (_G165_ and _G195_ above).

After we have concluded PR #3669, I am interested in doing a pull request for this bug. Would this be useful and possibly accepted?

Thanks,


Nick

[wcag-2.4.7]: https://w3.org/TR/WCAG21/#focus-visible
[qref-2.4.7]: https://w3.org/WAI/WCAG21/quickref/?showtechniques=247#focus-visible
[F78]: https://www.w3.org/WAI/WCAG21/Techniques/failures/F78.html
[G165]: https://www.w3.org/WAI/WCAG21/Techniques/general/G165.html
[G195]: https://www.w3.org/WAI/WCAG21/Techniques/general/G195.html
[C15]: https://www.w3.org/WAI/WCAG21/Techniques/css/C15.html
  "C15: Using CSS to change the presentation of a user interface component when it receives focus"

**Comment by Amit8527510735** [comment]
As observed the upload and file buttons are getting darker when the keyboard focus moves on the controls, but as discussed with internal SME, the keyboard focus indicator(Dotted line/solid line border) should be visible on both controls.

**Comment by corinagum** [comment]
note to dev: Based focus indicator on fluent design

**Comment by corinagum** [comment]
I'm adding this to our front-burner to hopefully be addressed in our next milestone.

**Comment by nfreear** [comment]
Hi @Amit8527510735,

Thank you for your response regarding the _subject matter expert's_ (SME's) advise (_I mis-read it a few days ago_!)

@corinagum ... If I can get PR ##3669 sorted quickly, I'm happy to contribute a pull request for this.

Thanks again,

Nick
